### ISSUE TITLE
[RFC] Distributed manual_seed API for N-D parallelism

### ISSUE THREAD

**Issue by wconstab** [pytorch/pytorch#140301]
### Background
In distributed training scenarios, RNG initialization matters for ensuring a correct model initialization, and in some cases also controlling random ops during training  (e.g. dropout).  For model initialization, the objective is to ensure that when different ranks initialize different shards of a layer (TP, FSDP) or different layers (PP), they get different values, and when different ranks initialize replicas of the same layer or the same shard (DDP or HSDP), they get the same value.

### DTensor RNG Tracker
Currently, DTensor supports RNG management.  There are some issues with this.
- It does not account for pipeline parallelism, and needs to be extended
- the design goals are over-ambitious (trying to enable matching the behavior with single-gpu models when this is not generally possible), leading to overly complex implementation
- the user-set RNG seeds are not (always) respected, and may be overwritten/ignored, depending on which RNG Tracker implementation DTensor selects (which is not user-visible, but depends on the degrees of parallelism used)

### Proposed API
`torch.distributed.manual_seed(seed, sharded_groups=[pp, tp, fsdp])`

An api like this could solve several needs:
- initialize ranks in 'sharded_groups' with different seeds by adding different offset values to the user-provided seed per rank, while seeding ranks of other groups with the same seed (to handle replica cases)
- additionally, `sharded_dims` arg would be accepted for device-mesh users,  only one used at a atime
- order of sharded groups specifies the order of incrementing the seed-offset, giving control to the user, and enabling things like trying to match the seeding behavior of megatron or some existing trainer

To enable this, we should change a few things about DTensor
- avoid overwriting user-provided seed.  Require the user to set the seed, but don't do it automatically
- don't broadcast the seed.  This can lead to collective hangs in some cases or just unexpected / wrong seeding behavior, since DTensor broadcasts to the whole world rather than knowing which groups are sharded.
- Use one RNG tracker consistently.  Avoid switching it depending on the parallelism.  Rely on this new API for goals like matching megatron RNG behavior.

**Comment by wanchaol** [comment]
> trying to enable matching the behavior with single-gpu models when this is not generally possible

I think there's definitely complications due to the way that the current PyTorch cuda eager mode behaves (i.e. threading to fill blocks), but I think the current behavior is already a single GPU behavior, it's just that it does not match exactly with the PyTorch cuda eager mode implementation as it's using simply philox seed + offset (this is similar to the situation of TorchInductor, where it could not match exactly with the PyTorch eager cuda implementation), cc @ezyang @Chillee

> torch.distributed.manual_seed(seed, sharded_groups=[pp, tp, fsdp])

There's actually a similar manual_seed API in the random module that serves similar purpose. https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_random.py#L51 Whatever we are implementing now, we should consider merge this with the new APIs proposed, and remove the collective calls inside this API.

My general feeling is that the automatic broadcasting the initial seed is quite critical to ensure single device semantic, as for users who don't want to call this `manual_seed` API they should be able to have a convergence guarantee (i.e. their world size is small so broadcasting once for the world is cheap enough). I also agree that if user know exactly what they are doing and want to control their own seeds, the DTensor RNG tracker should respect the seed that user already set. i.e. what we can do is that:
* if user calls `torch.distributed.manual_seed` already in their program then the DTensor RNG tracker simply do nothing
* If user don't call the manual_seed API offered by us, then DTensor RNG broadcast the seed lazily when hitting the first random op.

**Comment by wconstab** [comment]
> I think there's definitely complications due to the way that the current PyTorch cuda eager mode behaves (i.e. threading to fill blocks), but I think the current behavior is already a single GPU behavior, it's just that it does not match exactly with the PyTorch cuda eager mode implementation as it's using simply philox seed + offset (this is similar to the situation of TorchInductor, where it could not match exactly with the PyTorch eager cuda implementation), cc @ezyang @Chillee

I don't think the RNG situation for DTensor is the same as the one for inductor.  I'm not actually sure what the situation for inductor is, so maybe i'm wrong.  But the issue for DTensor is specific to two things:
* applies mainly to NN initialization (which i guess isn't commonly done inside of inductor)
* is specifically caused by tensor sharding

My point is that the only way you could initialize a layer with the same random values between single-gpu and sharded is if the original tensor shape is an exact multiple of the RNG pool size for the GPU you are using.  For the A100 i tested, this number is `884,736` so it is pretty unlikely IMO that any real models have all their layers a multiple of this size.  Absent this, the initialization values obtained from DTensor will not match the initialization values from eager, regardless of the seed management infra.  See [this note](https://docs.google.com/document/d/1Wa6cDQB2Imde4wqagZrVzChCvRMFnUQguhmWZpm2t2g/edit?tab=t.0#heading=h.5edk3sdj5uv) for more details.

I consider 2 goals for the distributed RNG machinery.  
1. get matching results between single-gpu and multi-gpu, so exact comparisons (logits, loss, etc) can be made between different parallelisms, and unit testing can be simplified
2. ensure correct distributed training - regardless of (1), ensure that different shards of a layer get different init values, by ensuring different seeds are used.  Give up on exactly matching single-gpu

I feel that goal 1 is impossible.  I also think goal 1 is 'all or nothing' - it doesn't make sense to try to "almost match" initialization values.  So if we agree that goal 1 is impossible, we shouldn't try to achieve it.

Goal 2 is critical.  But I think it is completely solved just by setting the initial seed for each rank.  I'm not convinced that forking the RNG for DTensor and micromanaging it for each operation adds value.

> My general feeling is that the automatic broadcasting the initial seed is quite critical to ensure single device semantic, as for users who don't want to call this manual_seed API they should be able to have a convergence guarantee (i.e. their world size is small so broadcasting once for the world is cheap enough). I also agree that if user know exactly what they are doing and want to control their own seeds, the DTensor RNG tracker should respect the seed that user already set. i.e. what we can do is similar to what you proposed above:

This point makes sense to me. I think it is debatable whether we should hide this detail from users or just document that users doing sharding need to ensure different seeds.   I'd be fine leaving the broadcast in place if we can fix the composability issues with it.
* it's not correct to broadcast to the whole world.  DTensor must know which dims are 'shard' vs 'replicate' before it does this broadcast.  Can this be determined automatically?  If not, we are actually introducing a bug here
* Disabling the broadcast when users set manual seed is a bit fragile.  I'm worried about a few cases- (a) users could set seeds via different APIs -- directly set cuda's seed, call torch.manual_seed() themselves on each rank, or use distributed.manual_seed() API.  It seems hard to let DTensor reliably know whether users set their seed or not.  Perhaps the best we can do is set a flag inside `distributed.manual_seed`, but we'd be missing some corner cases.

**Comment by XilunWu** [comment]
> get matching results between single-gpu and multi-gpu, so exact comparisons (logits, loss, etc) can be made between different parallelisms, and unit testing can be simplified

I agree this is impossible and this is neither something PyTorch guarantees. "The only expectation is that if one gives the seed sequences should stay the same every time code is run on the same HW/SW configuration" (https://github.com/pytorch/pytorch/issues/118735#issuecomment-1919488187).

> ensure correct distributed training - regardless of (1), ensure that different shards of a layer get different init values, by ensuring different seeds are used. Give up on exactly matching single-gpu

This is the main reason we introduced DTensor RNG and we believe the current implementation is the right approach from the perspective of DTensor (SPMD) instead of N-D parallelism composability. See below.

> Goal 2 is critical. But I think it is completely solved just by setting the initial seed for each rank. I'm not convinced that forking the RNG for DTensor and micromanaging it for each operation adds value.

The proposed seed setting solution assumes that the DTensor initialized over a mesh dimension is always Shard or Replicate (i.e. the mesh dimension is a Shard dimension or Replicate dimension). However, this is not generally true for DTensor: users can choose to shard one DTensor over a device mesh dim and replicate another DTensor over the same dim. This is why we believe using the same seed for all ranks and taking care of RNG offset for each DTensor random ops based on its sharding placements is the best way to achieve this duality.

> it's not correct to broadcast to the whole world. DTensor must know which dims are 'shard' vs 'replicate' before it does this broadcast. Can this be determined automatically? If not, we are actually introducing a bug here

I think the WORLD broadcast is correct. The real issue is DTensor assumes SPMD while PP is not actually SPMD where a WORLD broadcast leads to deadlock. DTensor doesn't need to know which dims are 'shard' vs 'replicate' because it sets the same seed on all ranks and adjusts RNG offset based on DTensor placements to achieve the "DTensor shards have different random numbers and DTensor replicas have the same random numbers" duality.

> Disabling the broadcast when users set manual seed is a bit fragile.

**Comment by wconstab** [comment]
> 1. if user has ever called torch.distributed.manual_seed before any RNG tracker initialization (the first DTensor random ops call), a RNG tracker will be initialized using the passed in seed and shard device mesh information. No broadcast will be ever performed. We can add a run_check bool option to this API to explicitly broadcast if users want to make sure the seed argument passed in on all ranks has the same value.

> 2. if user has not called torch.distributed.manual_seed, the RNG tracker will be initialized with a WORLD broadcast. This is the current behavior.

I think we still have a design gap here.  Consider 2 rank FSDP example.  (a) with manual_shard, each rank gets a different seed and dtensor skips broadcast.  But dtensor still manages 'offset'.  (b) without manual_shard, each rank has unknown seed and DTensor broadcasts rank0's state.  DTensor manages offset.  But case (a) and (b) will give different results, becuase for (b) the seeds on all ranks are the same while in (a) they are different.  They can't both be right!

> I think the WORLD broadcast is correct. The real issue is DTensor assumes SPMD while PP is not actually SPMD

I think I just object to making an assumption here.  We shouldn't assume SPMD unless told otherwise via a specific API.

**Comment by XilunWu** [comment]
> DTensor manages offset. But case (a) and (b) will give different results, becuase for (b) the seeds on all ranks are the same while in (a) they are different. They can't both be right!

My understanding is in case (a) where user calls the new API, they take full control of and are responsible for the RNG. DTensor shouldn't do offset modification in this case -- Wanchao suggests the RNG tracker do nothing including isolating the 2 RNG states. This is exactly the pytorch native RNG behavior except the new API sets up the seed.

(b) without manual_shard, DTensor takes care of the RNG.

Pardon me but I don't see why we must make (a) and (b) produce the same result.

**Comment by wconstab** [comment]
In case (a), if dtensor chooses the same offset for two ranks, that's because dtensor thinks those ranks are doing a replicated operation and should get same results. If the seed differs for those ranks due to manual seed and no broadcast, dtensor logic breaks down and those ranks no longer get same results.

**Comment by XilunWu** [comment]
Propose #141220 #141223 to handle the N-D parallelism RNG case. I believe we can call `torch.random` to set up the RNG seed. If we still think the distributed manual_seed API is needed, it can be implemented using .
